src/onnx/parse_if.cpp

```diff
@@ -65,13 +67,108 @@ struct parse_if : op_parser<parse_if>
+            {
+                MIGRAPHX_THROW("PARSE_IF: " + info.name +
+                               " then and else sub_grahps must have same output type! " +
+                               then_out_shapes.at(i).type_string() + " vs " +
+                               else_out_shapes.at(i).type_string());
+            }
+
+            if(not then_out_shapes.at(i).dynamic() && not else_out_shapes.at(i).dynamic())
+            {
+                auto then_lens = then_out_shapes.at(i).lens();
+                auto else_lens = else_out_shapes.at(i).lens();
+
+                // Throw error if both branches have zero output shapes. Not possible for static
+                // inputs
+                if(then_lens.empty() && else_lens.empty())
+                {
+                    throw_shapes();
+                }
+
+                auto handle_empty_branch = [](module_ref& mdl, const shape& out_shape) {
+                    auto outline_ins = mdl->add_outline(out_shape);
+                    mdl->replace_return({outline_ins});
+                    return out_shape.lens();
+                };
+
+                // Handle one empty branch by setting output identical to the other
+                // need to update the then_shape before we do further checks
+                if(then_lens.empty())
+                {
+                    then_lens = handle_empty_branch(then_mdl, else_out_shapes.at(i));
+                }
+                else if(else_lens.empty())
+                {
+                    else_lens = handle_empty_branch(else_mdl, then_out_shapes.at(i));
+                }
+
+                auto all_but_last_dims_equal = [](const std::vector<size_t>& lens_a,
+                                                  const std::vector<size_t>& lens_b) {
+                    if(lens_a.size() <= lens_b.size())
+                    {
+                        return std::equal(lens_a.begin(), lens_a.end(), lens_b.begin());
+                    }
+                    else
+                    {
+                        return std::equal(lens_b.begin(), lens_b.end(), lens_a.begin());
+                    }
+                };
+
+                // check equivalent length dims, and (x1,x2,.., xn, 1) == (x1,x2,..,xn)
+                int dim_delta = abs((static_cast<int>(then_lens.size() - else_lens.size())));
+
+                if(dim_delta == 1)
+                {
+                    // make sure dims are equivalent in static shapes
+                    if(not all_but_last_dims_equal(then_lens, else_lens))
+                    {
+                        throw_shapes();
+                    }
+
+                    auto last_then = then_lens.back();
+                    auto last_else = else_lens.back();
+
+                    auto unsqueeze_last_op = [](module_ref& mdl,
+                                                const std::vector<size_t>& out_shape) {
+                        auto convert_ins = mdl->add_instruction(
```

Comment:
You need to do `mdl->insert_instruction(std::prev(mdl->end()), ...)` to insert the instruction before the return.

Reply:
Ah is that why I'm seeing two returns and need to remove the one then?

Reply:
yep

Reply:
Ah, again thank you for the insight. Thought it was a simpler than it is. This is done.  I guess this has to do with how we handle our returns at the end of each graph?